mavros/src/plugins/setpoint_trajectory.cpp

```diff
@@ -0,0 +1,162 @@
+				pose_msg.pose.position.y = req->points[i].transforms[0].translation.y;
+				pose_msg.pose.position.z = req->points[i].transforms[0].translation.z;
+				pose_msg.pose.orientation.w = req->points[i].transforms[0].rotation.w;
+				pose_msg.pose.orientation.x = req->points[i].transforms[0].rotation.x;
+				pose_msg.pose.orientation.y = req->points[i].transforms[0].rotation.y;
+				pose_msg.pose.orientation.z = req->points[i].transforms[0].rotation.z;
+				msg.poses.emplace_back(pose_msg);
+			}
+		}
+		desired_pub.publish(msg);
+	}
+
+	/* -*- callbacks -*- */
+
+	void local_cb(const trajectory_msgs::MultiDOFJointTrajectory::ConstPtr &req)
+	{
+		trajectory_target_msg = req;
+		refstart_time = ros::Time::now();
+		publish_path(req);
+	}
+
+	void reference_cb(const ros::TimerEvent &event)
+	{
+		if(trajectory_target_msg) {
+			ros::Duration curr_time_from_start;
+			curr_time_from_start = ros::Time::now() - refstart_time;
+		
+			for(size_t i = 0; i < trajectory_target_msg->points.size(); i++){
+
+				Eigen::Vector3d position, velocity, af;
+				Eigen::Quaterniond attitude;
+				float yaw, yaw_rate;
+				trajectory_msgs::MultiDOFJointTrajectoryPoint pt = trajectory_target_msg->points[i];
+				uint16_t type_mask;
+
+				if(pt.time_from_start.toSec() >= curr_time_from_start.toSec() ) { //TODO: Better logic to handle this case?
+					if(!pt.transforms.empty()){
+					position << pt.transforms[0].translation.x, pt.transforms[0].translation.y, pt.transforms[0].translation.z;
+					attitude = Eigen::Quaterniond(pt.transforms[0].rotation.w, pt.transforms[0].rotation.x, pt.transforms[0].rotation.y, pt.transforms[0].rotation.z);
+					} else {
+						type_mask = type_mask || mavros_msgs::PositionTarget::IGNORE_PX || mavros_msgs::PositionTarget::IGNORE_PY || mavros_msgs::PositionTarget::IGNORE_PZ;
+					}
+
+					if(!pt.velocities.empty()) velocity << pt.velocities[0].linear.x, pt.velocities[0].linear.y, pt.velocities[0].linear.z;
+					else type_mask = type_mask || mavros_msgs::PositionTarget::IGNORE_VX || mavros_msgs::PositionTarget::IGNORE_VY || mavros_msgs::PositionTarget::IGNORE_VZ;
+					
+					if(!pt.accelerations.empty()) af << pt.accelerations[0].linear.x, pt.accelerations[0].linear.y, pt.accelerations[0].linear.z;
+					else type_mask = type_mask || mavros_msgs::PositionTarget::IGNORE_VX || mavros_msgs::PositionTarget::IGNORE_VY || mavros_msgs::PositionTarget::IGNORE_VZ;
+
+					// Transform frame ENU->NED
+					position = ftf::transform_frame_enu_ned(position);
+					velocity = ftf::transform_frame_enu_ned(velocity);
+					af = ftf::transform_frame_enu_ned(af);
+					yaw = ftf::quaternion_get_yaw(attitude);
+
+					set_position_target_local_ned(
+								trajectory_target_msg->header.stamp.toNSec() / 1000000,
+								1,
+								0,
+								position,
+								velocity,
+								af,
+								yaw, 0);
+					if(i == trajectory_target_msg->points.size()-1)  trajectory_target_msg.reset(); //End of trajectory
+					break;
```

Comment:
I think here better to use iterators and compare against `v::cend()`.

Reply:
I decided it is better to return the function if the trajectory is published and delete the message when it makes it to the end of the iterator